orchard_group_hash.py

```diff
@@ -0,0 +1,167 @@
+def sxor(s1,s2):
+    return bytes([a ^ b for a,b in zip(s1,s2)])
+
+def expand_message_xmd(msg, dst, len_in_bytes):
+    assert len(dst) <= 255
+
+    b_in_bytes = 64 # hash function output size
+    r_in_bytes = 128
+
+    ell = cldiv(len_in_bytes, b_in_bytes)
+
+    assert ell <= 255
+
+    dst_prime = dst + i2beosp(8, len(dst))
+    z_pad = b"\x00" * r_in_bytes
+    l_i_b_str = i2beosp(16, len_in_bytes)
+    msg_prime = z_pad + msg + l_i_b_str + i2beosp(8, 0) + dst_prime
+
+    b = []
+
+    b0_ctx = blake2b(digest_size=b_in_bytes, person=i2beosp(128,0))
+    b0_ctx.update(msg_prime)
+    b.append(b0_ctx.digest())
+    assert len(b[0]) == b_in_bytes
+
+    b1_ctx = blake2b(digest_size=b_in_bytes, person=i2beosp(128,0))
+    b1_ctx.update(b[0] + i2beosp(8, 1) + dst_prime)
+    b.append(b1_ctx.digest())
+    assert len(b[1]) == b_in_bytes
+
+    for i in range(2, ell + 1):
+        bi_input = sxor(b[0], b[i-1])
+
+        assert len(bi_input) == b_in_bytes
+
+        bi_input += i2beosp(8, i) + dst_prime
+
+        bi_ctx = blake2b(digest_size=b_in_bytes, person=i2beosp(128,0))
+        bi_ctx.update(bi_input)
+
+        b.append(bi_ctx.digest())
+        assert len(b[i]) == b_in_bytes
+
+    return b''.join(b[1:])[0:len_in_bytes]
+
+def hash_to_field(msg, dst):
+    k = 256
+    count = 2
+    m = 1
+
+    L = cldiv(math.ceil(math.log2(p)) + k, 8)
+    assert L == 512/8
+
+    len_in_bytes = count * m * L
+    uniform_bytes = expand_message_xmd(msg, dst, len_in_bytes)
+
+    elements = []
+    for i in range(0, count):
+        for j in range(0, m):
+            elm_offset = L * (j + i * m)
+            tv = uniform_bytes[elm_offset:elm_offset+L]
+            elements.append(Fp(beos2ip(tv), False))
+
+    assert len(elements) == count
+
+    return elements
+
+def map_to_curve_simple_swu(u):
+    # The notation below follows Appendix F.2 of the Internet Draft
+    zero = Fp(0)
+    assert zero.inv() == Fp(0)
+
+    A = PALLAS_ISO_A
+    B = PALLAS_ISO_B
+    Z = Fp(-13, False)
+    c1 = -B / A
+    c2 = Fp(-1) / Z
+
+    tv1 = Z * u.exp(2)
+    tv2 = tv1.exp(2)
+    x1 = tv1 + tv2
+
+    x1 = x1.inv()
+    e1 = x1 == Fp(0)
+    x1 = x1 + Fp(1)
+
+    if e1:
+        x1 = c2
+    else:
+        x1 = x1
+
+    x1 = x1 * c1      # x1 = (-B / A) * (1 + (1 / (Z^2 * u^4 + Z * u^2)))
+    gx1 = x1.exp(2)
+    gx1 = gx1 + A
+    gx1 = gx1 * x1
+    gx1 = gx1 + B             # gx1 = g(x1) = x1^3 + A * x1 + B
+    x2 = tv1 * x1            # x2 = Z * u^2 * x1
+    tv2 = tv1 * tv2
+    gx2 = gx1 * tv2           # gx2 = (Z * u^2)^3 * gx1
+
+    e2 = (gx1.sqrt() is not None)
+
+    x = x1 if e2 else x2    # If is_square(gx1), x = x1, else x = x2
+    yy = gx1 if e2 else gx2  # If is_square(gx1), yy = gx1, else yy = gx2
+    y = yy.sqrt()
+
+    e3 = u.sgn0() == y.sgn0()
+
+    y = y if e3 else -y  #y = CMOV(-y, y, e3)
+
+    return orchard_iso_pallas.Point(x, y).iso_map()
```

Comment:
The algorithm in Appendix F.2 does not include isogeny map in its `map_to_curve_simple_swu`. Consider refactoring so that this more closely matches the draft.

Reply:
This is done in my first commit (322aff1777f8ccd74bbea85eb05d65d845a5827e).